File: src/resolver/greedy.cpp
#include "greedy.hpp"
#include "../core/dependency.hpp"
#include <algorithm>
#include <sstream>

using namespace std;

GreedyResolver::GreedyResolver() {}

pair<DependencyGraph, vector<string>> GreedyResolver::buildGraphWithErrors(
    const vector<string>& requested_packages,
    const vector<shared_ptr<Source>>& sources) {
    
    DependencyGraph graph;
    vector<string> not_found;
    
    // Queue for BFS traversal
    vector<string> queue = requested_packages;
    map<string, bool> processed;
    
    // Prefetch initial packages in parallel
    for (const auto& source : sources) {
        source->prefetch(requested_packages);
    }
    
    while (!queue.empty()) {
        // Batch process: collect unprocessed packages
        vector<string> batch;
        while (!queue.empty() && batch.size() < 10) {
            string pkg = queue.back();
            queue.pop_back();
            if (!processed[pkg]) {
                batch.push_back(pkg);
                processed[pkg] = true;
            }
        }
        
        if (batch.empty()) continue;
        
        // Prefetch batch dependencies
        for (const auto& source : sources) {
            source->prefetch(batch);
        }
        
        // Process batch
        for (const auto& package_name : batch) {
            // Find source for this package
            shared_ptr<Source> source = nullptr;
            for (const auto& s : sources) {
                if (s->packageExists(package_name)) {
                    source = s;
                    break;
                }
            }
            
            if (!source) {
                not_found.push_back(package_name);
                continue;
            }
            
            // Get latest version (greedy: try latest first)
            auto package_opt = source->fetchLatest(package_name);
            if (!package_opt.has_value()) {
                not_found.push_back(package_name);
                continue;
            }
            
            Package package = package_opt.value();
            graph.addPackage(package);
            
            // Collect dependencies for next batch prefetch
            vector<string> deps_to_prefetch;
            for (const auto& dep : package.getDependencies()) {
                string dep_name = dep.getName();
                graph.addDependency(package_name, dep_name);
                
                if (!processed[dep_name]) {
                    queue.push_back(dep_name);
                    deps_to_prefetch.push_back(dep_name);
                }
            }
            
            // Prefetch dependencies
            if (!deps_to_prefetch.empty()) {
                for (const auto& s : sources) {
                    s->prefetch(deps_to_prefetch);
                }
            }
        }
    }
    
    return {graph, not_found};
}

vector<VersionConstraint> GreedyResolver::getConstraints(const string& package,
                                                        const DependencyGraph& graph) {
    vector<VersionConstraint> constraints;
    
    // Get all dependents (packages that depend on this one)
    auto dependents = graph.getDependents(package);
    
    for (const auto& dependent : dependents) {
        auto dep_package_opt = graph.getPackage(dependent);
        if (!dep_package_opt.has_value()) {
            continue;
        }
        
        Package dep_package = dep_package_opt.value();
        
        // Find constraint for this package in dependent's dependencies
        for (const auto& dep : dep_package.getDependencies()) {
            if (dep.getName() == package) {
                auto dep_constraints = dep.getConstraints();
                constraints.insert(constraints.end(), 
                                 dep_constraints.begin(), 
                                 dep_constraints.end());
            }
        }
    }
    
    return constraints;
}

int GreedyResolver::scoreVersion(const Version& version,
                                 [[maybe_unused]] const vector<VersionConstraint>& constraints,
                                 const map<string, string>& selected) {
    int score = 0;
    
    // Prefer stable versions
    if (version.isStable()) {
        score += 1000;
    }
    
    // Prefer newer versions (higher major/minor/patch)
    score += version.getMajor() * 100;
    score += version.getMinor() * 10;
    score += version.getPatch();
    
    // Check if this version is already selected (reuse optimization)
    for (const auto& pair : selected) {
        try {
            Version selected_version(pair.second);
            if (selected_version == version) {
                score += 500; // Bonus for reusing
            }
        } catch (...) {
            // Ignore
        }
    }
    
    return score;
}

optional<string> GreedyResolver::selectBestVersion([[maybe_unused]] const string& package,
                                                  const vector<string>& available_versions,
                                                  const vector<VersionConstraint>& constraints,
                                                  [[maybe_unused]] const vector<shared_ptr<Source>>& sources) {
    vector<pair<Version, string>> valid_versions; // (version_obj, version_str)
    
    // Filter versions by constraints
    for (const auto& version_str : available_versions) {
        try {
            Version version(version_str);
            
            // Check if version satisfies all constraints
            bool satisfies_all = true;
            for (const auto& constraint : constraints) {
                if (!constraint.satisfies(version)) {
                    satisfies_all = false;
                    break;
                }
            }
            
            if (satisfies_all) {
                valid_versions.push_back({version, version_str});
            }
        } catch (...) {
            // Skip invalid versions
        }
    }
    
    if (valid_versions.empty()) {
        return nullopt;
    }
    
    // Sort by score (greedy heuristic)
    sort(valid_versions.begin(), valid_versions.end(),
         [this, &constraints](const pair<Version, string>& a,
                              const pair<Version, string>& b) {
             int score_a = scoreVersion(a.first, constraints, selected_versions);
             int score_b = scoreVersion(b.first, constraints, selected_versions);
             return score_a > score_b; // Higher score is better
         });
    
    return valid_versions[0].second;
}

GreedyResult GreedyResolver::resolve(const vector<string>& requested_packages,
                                    const vector<shared_ptr<Source>>& sources) {
    selected_versions.clear();
    GreedyResult result;
    result.success = false;
    
    // Build dependency graph
    auto [graph, not_found] = buildGraphWithErrors(requested_packages, sources);
    result.not_found_packages = not_found;
    
    // If any requested package was not found, fail immediately
    for (const auto& pkg : requested_packages) {
        bool found = false;
        for (const auto& nf : not_found) {
            if (nf == pkg) {
                found = true;
                break;
            }
        }
        if (found) {
            result.conflict_package = pkg;
            result.conflict_reason = "Package not found in any source: " + pkg;
            return result;
        }
    }
    
    // Check for cycles
    if (graph.hasCycle()) {
        result.conflict_reason = "Circular dependency detected";
        auto cycle = graph.getCycle();
        if (!cycle.empty()) {
            result.conflict_package = cycle[0];
        }
        return result;
    }
    
    // Topological sort to get resolution order
    vector<string> resolution_order = graph.topologicalSort();
    
    // Resolve each package in order
    for (const auto& package_name : resolution_order) {
        // Find source for this package
        shared_ptr<Source> source = nullptr;
        for (const auto& s : sources) {
            if (s->packageExists(package_name)) {
                source = s;
                break;
            }
        }
        
        if (!source) {
            result.conflict_package = package_name;
            result.conflict_reason = "Package not found in any source: " + package_name;
            return result;
        }
        
        // Get available versions
        vector<string> available_versions = source->getAvailableVersions(package_name);
        if (available_versions.empty()) {
            result.conflict_package = package_name;
            result.conflict_reason = "No versions available for: " + package_name;
            return result;
        }
        
        // Get constraints from dependents
        vector<VersionConstraint> constraints = getConstraints(package_name, graph);
        
        // Select best version using greedy heuristics
        auto best_version = selectBestVersion(package_name, available_versions, 
                                             constraints, sources);
        
        if (!best_version.has_value()) {
            result.conflict_package = package_name;
            result.conflict_reason = "No version satisfies constraints for: " + package_name;
            return result;
        }
        
        selected_versions[package_name] = best_version.value();
    }
    
    result.success = true;
    result.selected_versions = selected_versions;
    return result;
}

File: dpm/resolver/greedy.py
"""
greedy.py - greedy resolver - fast but might not find solution if there are conflicts
basically just picks the latest version that works and hopes for the best
"""

from typing import List, Dict, Optional, Tuple
from dataclasses import dataclass
from .graph import DependencyGraph
from ..core.package import Package
from ..core.version import Version, VersionConstraint
from ..core.dependency import Dependency
from ..sources.source import Source


@dataclass
class GreedyResult:
    """result from greedy resolution attempt"""
    success: bool
    selected_versions: Dict[str, str]  # what versions we picked
    conflict_package: str = ""  # which package caused issues
    conflict_reason: str = ""
    not_found_packages: List[str] = None  # packages we couldn't find anywhere
    conflict_details: Dict[str, List[str]] = None  # package -> list of conflicting requirements
    
    def __post_init__(self):
        if self.not_found_packages is None:
            self.not_found_packages = []
        if self.conflict_details is None:
            self.conflict_details = {}


class GreedyResolver:
    """greedy resolver - fast but might not find solution if there are conflicts"""
    
    def __init__(self):
        self.selected_versions: Dict[str, str] = {}
    
    def resolve(self, requested_packages: List[str], sources: List[Source]) -> GreedyResult:
        """try to resolve using greedy approach"""
        self.selected_versions.clear()
        result = GreedyResult(success=False, selected_versions={})
        
        # build dependency graph
        graph, not_found = self._build_graph_with_errors(requested_packages, sources)
        result.not_found_packages = not_found
        
        # if any requested package was not found, fail immediately
        for pkg in requested_packages:
            if pkg in not_found:
                result.conflict_package = pkg
                result.conflict_reason = f"Package not found in any source: {pkg}"
                return result
        
        # check for cycles
        if graph.has_cycle():
            result.conflict_reason = "Circular dependency detected"
            cycle = graph.get_cycle()
            if cycle:
                result.conflict_package = cycle[0]
            return result
        
        # topological sort to get resolution order
        resolution_order = graph.topological_sort()
        if not resolution_order:
            result.conflict_reason = "Failed to determine resolution order"
            return result
        
        # resolve each package in order
        for package_name in resolution_order:
            # find source for this package
            source = None
            for s in sources:
                if s.package_exists(package_name):
                    source = s
                    break
            
            if not source:
                result.conflict_package = package_name
                result.conflict_reason = f"Package not found in any source: {package_name}"
                return result
            
            # get available versions
            available_versions = source.get_available_versions(package_name)
            if not available_versions:
                result.conflict_package = package_name
                result.conflict_reason = f"No versions available for: {package_name}"
                return result
            
            # get constraints from dependents
            constraints = self._get_constraints(package_name, graph)
            
            # select best version using greedy heuristics
            best_version = self._select_best_version(
                package_name, available_versions, constraints, sources
            )
            
            if not best_version:
                result.conflict_package = package_name
                result.conflict_reason = f"No version satisfies constraints for: {package_name}"
                return result
            
            self.selected_versions[package_name] = best_version
        
        result.success = True
        result.selected_versions = self.selected_versions.copy()
        return result
    
    def _build_graph_with_errors(self, requested_packages: List[str], 
                                 sources: List[Source]) -> Tuple[DependencyGraph, List[str]]:
        """builds the dependency graph and tracks which packages we couldn't find"""
        graph = DependencyGraph()
        not_found = []
        
        # queue for bfs traversal
        queue = requested_packages.copy()
        processed = set()
        
        # prefetch initial packages in parallel
        for source in sources:
            source.prefetch(requested_packages)
        
        while queue:
            # batch process: collect unprocessed packages
            batch = []
            while queue and len(batch) < 10:
                pkg = queue.pop()
                if pkg not in processed:
                    batch.append(pkg)
                    processed.add(pkg)
            
            if not batch:
                continue
            
            # prefetch batch dependencies
            for source in sources:
                source.prefetch(batch)
            
            # process batch
            for package_name in batch:
                # find source for this package
                source = None
                for s in sources:
                    if s.package_exists(package_name):
                        source = s
                        break
                
                if not source:
                    not_found.append(package_name)
                    continue
                
                # get latest version (greedy: try latest first)
                package = source.fetch_latest(package_name)
                if not package:
                    not_found.append(package_name)
                    continue
                
                graph.add_package(package)
                
                # collect dependencies for next batch prefetch
                deps_to_prefetch = []
                for dep in package.dependencies:
                    dep_name = dep.name
                    graph.add_dependency(package_name, dep_name)
                    
                    if dep_name not in processed:
                        queue.append(dep_name)
                        deps_to_prefetch.append(dep_name)
                
                # prefetch dependencies
                if deps_to_prefetch:
                    for s in sources:
                        s.prefetch(deps_to_prefetch)
        
        return graph, not_found
    
    def _get_constraints(self, package: str, graph: DependencyGraph) -> List[VersionConstraint]:
        """gets all the constraints on a package from things that depend on it"""
        constraints = []
        
        # get all dependents (packages that depend on this one)
        dependents = graph.get_dependents(package)
        
        for dependent in dependents:
            dep_package = graph.get_package(dependent)
            if not dep_package:
                continue
            
            # find constraint for this package in dependent's dependencies
            for dep in dep_package.dependencies:
                if dep.name == package:
                    constraints.extend(dep.constraints)
        
        return constraints
    
    def _score_version(self, version: Version, constraints: List[VersionConstraint],
                      selected: Dict[str, str]) -> int:
        """scores a version - higher is better"""
        score = 0
        
        # prefer stable versions
        if version.is_stable():
            score += 1000
        
        # prefer newer versions (higher major/minor/patch)
        score += version.major * 100
        score += version.minor * 10
        score += version.patch
        
        # check if this version is already selected (reuse optimization)
        for selected_version_str in selected.values():
            try:
                selected_version = Version(selected_version_str)
                if selected_version == version:
                    score += 500  # bonus for reusing
            except ValueError:
                pass
        
        return score
    
    def _select_best_version(self, package: str, available_versions: List[str],
                           constraints: List[VersionConstraint],
                           sources: List[Source]) -> Optional[str]:
        """picks the best version based on our heuristics
        prefers: latest stable > latest prerelease > already selected
        """
        valid_versions = []  # (version_obj, version_str)
        
        # filter versions by constraints
        for version_str in available_versions:
            try:
                version = Version(version_str)
                
                # check if version satisfies all constraints
                satisfies_all = True
                for constraint in constraints:
                    if not constraint.satisfies(version):
                        satisfies_all = False
                        break
                
                if satisfies_all:
                    valid_versions.append((version, version_str))
            except ValueError:
                # skip invalid versions
                continue
        
        if not valid_versions:
            return None
        
        # sort by score (greedy heuristic)
        valid_versions.sort(
            key=lambda x: self._score_version(x[0], constraints, self.selected_versions),
            reverse=True
        )
        
        return valid_versions[0][1]

